### PR TITLE
Cloudmigration: swagger for the cloudmigration api

### PR DESCRIPTION
**Pull request by xlson**
**What is this feature?**

Swagger api description for the cloudmigration api.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by papagian — COMMENTED** [review]
The OpenAPI changes are missing: you have follow [these instructions](https://github.com/grafana/grafana/blob/main/pkg/api/README.md#openapi-generation) to regenerate the specs and push the changes

**Comment by xlson** [comment]
> The OpenAPI changes are missing: you have follow [these instructions](https://github.com/grafana/grafana/blob/main/pkg/api/README.md#openapi-generation) to regenerate the specs and push the changes

Thanks @papagian! I'll fix that.

**Comment by xlson** [comment]
@papagian Do you think you could help me figure out what I'm doing wrong when I try to define the swagger api? I'm getting some errors when I try to re-generate the swagger definition and I fixed the ones I was able to figure out but there's some left.

```bash
$ make swagger-clean && make openapi3-gen
rm -f public/api-spec.json public/api-merged.json public/openapi3.json
generate go files
go run ./pkg/build/wire/cmd/wire/main.go gen -tags "oss" ./pkg/server
wire: github.com/grafana/grafana/pkg/server: wrote /home/leo/code/work/grafana/pkg/server/wire_gen.go
re-generating swagger for OSS
rm -f public/api-spec.json
SWAGGER_GENERATE_EXTENSION=false /home/leo/code/go/bin/swagger-v0.30.2 generate spec -q -m -w pkg/server -o public/api-spec.json \
-x "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions" \
-x "github.com/prometheus/alertmanager" \
-i pkg/api/swagger_tags.json \
--exclude-tag=alpha \
--exclude-tag=enterprise
skipping re-generating swagger for enterprise: not enabled
# known conflicts DsPermissionType, AddApiKeyCommand, Json, Duration (identical models referenced by both specs)
/home/leo/code/go/bin/swagger-v0.30.2 mixin -q public/api-spec.json public/api-enterprise-spec.json pkg/services/ngalert/api/tooling/api.json --ignore-conflicts -o public/api-merged.json
/home/leo/code/go/bin/swagger-v0.30.2 validate --skip-warnings public/api-merged.json
2024/03/28 15:12:58 
The swagger spec at "public/api-merged.json" showed up some valid but possibly unwanted constructs.

The swagger spec at "public/api-merged.json" is invalid against swagger specification 2.0.
See errors below:
- type in responses.cloudMigrationRunResponse.headers is required
- responses.cloudMigrationRunResponse.headers.result.type in body is required
- "migrations" is defined 2 times
- path param "{id}" has no parameter definition

make: *** [Makefile:80: swagger-validate] Error 1

```

File: pkg/services/cloudmigration/api/api.go [hunk 1]
```diff
@@ -51,6 +51,15 @@ func (cma *CloudMigrationAPI) registerEndpoints() {
 	}, middleware.ReqGrafanaAdmin)
 }
 
+// swagger:route POST /cloudmigration/token migrations
```
Comment (papagian) [inline pkg/services/cloudmigration/api/api.go:54]: need to define an operation id
```suggestion
// swagger:route POST /cloudmigration/token migrations  createCloudMigrationToken
```

File: pkg/services/cloudmigration/api/api.go [hunk 2]
```diff
@@ -66,14 +75,36 @@ func (cma *CloudMigrationAPI) CreateToken(c *contextmodel.ReqContext) response.R
 	return response.JSON(http.StatusOK, cloudmigration.CreateAccessTokenResponseDTO(resp))
 }
 
+// swagger:route GET /cloudmigration/migration migrations
```
Comment (papagian) [inline pkg/services/cloudmigration/api/api.go:78]:
```suggestion
// swagger:route GET /cloudmigration/migration migrations getMigrationList
```

File: pkg/services/cloudmigration/model.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ type MigratedResource struct {
 	Result MigratedResourceResult `json:"result"`
 }
 
+// swagger:response cloudMigrationRunResponse
```
Comment (papagian) [inline pkg/services/cloudmigration/model.go:40]:
```suggestion
```

File: pkg/services/cloudmigration/model.go [hunk 4]
```diff
@@ -47,17 +48,29 @@ type CloudMigrationRun struct {
 	Finished          time.Time          `json:"finished"`
 }
 
+// swagger:response cloudMigrationRunListResponse
```
Comment (papagian) [inline pkg/services/cloudmigration/model.go:51]:
```suggestion
```

File: pkg/services/cloudmigration/model.go [hunk 5]
```diff
@@ -47,17 +48,29 @@ type CloudMigrationRun struct {
...
+type CloudMigrationRunList struct {
+	Runs []CloudMigrationRun `json:"runs"`
+}
+
+// swagger:parameters createMigration
 type CloudMigrationRequest struct {
 	AuthToken string `json:"authToken"`
 }
 
+// swagger:response cloudMigrationResponse
```
Comment (papagian) [inline pkg/services/cloudmigration/model.go:61]:
```suggestion
```

File: pkg/services/cloudmigration/model.go [hunk 6]
```diff
@@ -47,17 +48,29 @@ type CloudMigrationRun struct {
...
 type CloudMigrationResponse struct {
 	ID      int64     `json:"id"`
 	Stack   string    `json:"stack"`
 	Created time.Time `json:"created"`
 	Updated time.Time `json:"updated"`
 }
 
+// swagger:response cloudMigrationListResponse
```
Comment (papagian) [inline pkg/services/cloudmigration/model.go:69]:
```suggestion
```

File: pkg/services/cloudmigration/model.go [hunk 7]
```diff
@@ -76,6 +89,7 @@ type MigrateDatasourcesResponseDTO struct {
 	DatasourcesMigrated int `json:"datasourcesMigrated"`
 }
 
+// swagger:response cloudMigrationCreateTokenResponse
```
Comment (papagian) [inline pkg/services/cloudmigration/model.go:92]:
```suggestion
```

File: pkg/services/cloudmigration/api/api.go [hunk 8]
```diff
@@ -114,18 +190,61 @@ func (cma *CloudMigrationAPI) GetMigrationRun(c *contextmodel.ReqContext) respon
...
-	return response.JSON(http.StatusOK, migrationStatus)
+
+	runList := cloudmigration.CloudMigrationRunList{Runs: migrationStatus}
+	return response.JSON(http.StatusOK, runList)
+}
+
+// swagger:parameters getCloudMigrationRunList
+type GetCloudMigrationRunList struct {
+	ID int64 `json:"id"`
 }
```
Comment (papagian) [inline pkg/services/cloudmigration/api/api.go:228]:
```suggestion
type GetCloudMigrationRunList struct {
    // ID of an migration
	//
	// in: path
	ID int64 `json:"id"`
}
```

Comment (papagian) [inline pkg/services/cloudmigration/api/api.go:228]:
```suggestion
// swagger:parameters getCloudMigrationRunList
type GetCloudMigrationRunList struct {
    // ID of an migration
	//
	// in: path
	ID int64 `json:"id"`
}
```

**Review by papagian — COMMENTED** [review]
for some reason responses should be defined as follows:
```
// swagger:response cloudMigrationRunResponse
type CloudMigrationRunResponse struct {
	// in: body
	Body cloudmigration.CloudMigrationRun
}

// swagger:response cloudMigrationListResponse
type CloudMigrationListResponse struct {
	// in: body
	Body cloudmigration.CloudMigrationListResponse
}

// swagger:response cloudMigrationResponse
type CloudMigrationResponse struct {
	// in: body
	Body cloudmigration.CloudMigrationResponse
}

// swagger:response cloudMigrationRunListResponse
type CloudMigrationRunListResponse struct {
	// in: body
	Body cloudmigration.CloudMigrationRunList
}

// swagger:response cloudMigrationCreateTokenResponse
type CloudMigrationCreateTokenResponse struct {
	// in: body
	Body cloudmigration.CreateAccessTokenResponseDTO
}
```

**Review by joshhunt — APPROVED** [review]
👍  excellent stuff!

**Pull request merged**